File: server/src/main/scala/com/lightbend/modelServer/flinkprocessors/SpeculativeProcessor.scala
/*
 * Copyright (C) 2017  Lightbend
 *
 * This file is part of flink-ModelServing
 *
 * flink-ModelServing is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

package com.lightbend.modelServer.flinkprocessors

import com.lightbend.modelServer.model._
import com.lightbend.modelServer.model.speculative.{CurrentProcessing, ServingResponse, SpeculativeExecutionStats, SpeculativeServiceRequest}
import com.lightbend.modelServer.processor.VotingDecider
import com.lightbend.speculative.speculativedescriptor.SpeculativeDescriptor
import org.apache.flink.api.common.state.{ValueState, ValueStateDescriptor}
import org.apache.flink.api.scala.createTypeInformation
import org.apache.flink.configuration.Configuration
import org.apache.flink.streaming.api.functions.co.CoProcessFunction
import org.apache.flink.util.Collector

import scala.collection.mutable.ListBuffer

object SpeculativeProcessor {
  val SERVERTIMEOUT = 150l
  def apply() = new SpeculativeProcessor
}

class SpeculativeProcessor extends CoProcessFunction[SpeculativeServiceRequest, SpeculativeDescriptor, ServingResult] {

  // The managed keyed state see https://ci.apache.org/projects/flink/flink-docs-release-1.4/dev/stream/state.html
  var specualtiveState: ValueState[SpeculativeExecutionStats] = _
  val speculativeDecider = VotingDecider
  import SpeculativeProcessor._
  var speculativeTimeout: ValueState[Long] = _

  // Because we are explicitely using unique GUIDs, we do not need to bring this to the key scope
  // Using this in the keygroup scope is good enough.
  val currentProcessing = collection.mutable.Map[String, CurrentProcessing]()

  override def open(parameters: Configuration): Unit = {
    val speculativeModelDesc = new ValueStateDescriptor[SpeculativeExecutionStats](
      "currentSpeculativeModel",   // state name
      createTypeInformation[SpeculativeExecutionStats]) // type information
    speculativeModelDesc.setQueryable("speculativeState")
    specualtiveState = getRuntimeContext.getState(speculativeModelDesc)
    val speculativeTmoutDesc = new ValueStateDescriptor[Long] (
      "Timeout",
      createTypeInformation[Long])
    speculativeTimeout = getRuntimeContext.getState(speculativeTmoutDesc)
  }

  override def processElement2(model: SpeculativeDescriptor, ctx: CoProcessFunction[SpeculativeServiceRequest, SpeculativeDescriptor, ServingResult]#Context, out: Collector[ServingResult]): Unit = {

    speculativeTimeout.update((model.tmout))
    if(specualtiveState.value() == null) specualtiveState.update(SpeculativeExecutionStats(model.datatype, speculativeDecider.getClass.getName, model.tmout))
  }

  override def processElement1(record: SpeculativeServiceRequest, ctx: CoProcessFunction[SpeculativeServiceRequest, SpeculativeDescriptor, ServingResult]#Context, out: Collector[ServingResult]): Unit = {

    if(specualtiveState.value == null) specualtiveState.update(SpeculativeExecutionStats(record.dataType, speculativeDecider.getClass.getName, SERVERTIMEOUT))
    if(speculativeTimeout.value() == 0) speculativeTimeout.update(SERVERTIMEOUT)

    // This will be invoked in 3 different use cases - process accordingly
    record.result match {
      case Some(result) => // This is either intermediate result from Model Processor or result from Router (no models)
        record.models match {
          case m if (m == 0) => // from the router - just forward
            out.collect(result)
          case _ => // Intermediate result
            currentProcessing.get(record.GUID) match {
              case Some(processingResults) =>
                // We are still waiting for this GUID
                val current = CurrentProcessing(processingResults.dataType, processingResults.models, processingResults.start, processingResults.results += ServingResponse(record.GUID, result))
                current.results.size match {
                  case size if (size >= current.models) => processResult(record.GUID, current, out) // We are done
                  case _ => val _ = currentProcessing += (record.GUID -> current) // Keep going
                }
              case _ => // Timed out
            }
        }
      case _ => // This is start collection
        // Add to the watch list
        val currentTime = System.currentTimeMillis()
        currentProcessing += (record.GUID -> CurrentProcessing(record.dataType, record.models, currentTime, new ListBuffer[ServingResponse]()))
        // Schedule timeout
        ctx.timerService.registerProcessingTimeTimer(currentTime + speculativeTimeout.value())
    }
  }

  override def onTimer(timestamp: Long, ctx: CoProcessFunction[SpeculativeServiceRequest, SpeculativeDescriptor, ServingResult]#OnTimerContext, out: Collector[ServingResult]): Unit = {
    // Complete timed out requests
    val start = System.currentTimeMillis() - speculativeTimeout.value
    currentProcessing.foreach(kv =>
      kv._2.start match {
        case t if (t < start) => // Timed out
          println(s"Timed out ${kv._1} at ${System.currentTimeMillis()} started at ${kv._2.start}")
          processResult(kv._1, kv._2, out)
        case _ => // keep going
      }
    )
  }

  // Complete speculative execution
  private def processResult(GUID : String, results: CurrentProcessing, out: Collector[ServingResult]) : Unit = {
    // Run it through decider
    val servingResult = speculativeDecider.decideResult(results).asInstanceOf[ServingResult]
    // Return valye
    out.collect(servingResult)
    // Update state
    if(servingResult.processed) {
      val state = specualtiveState.value().incrementUsage(servingResult.duration)
      specualtiveState.update(state)
    }
    // remove state
    val _ = currentProcessing -= GUID
  }
}

File: server/src/main/java/com/lightbend/java/modelServer/flinkprocessors/SpeculativeProcessor.java
package com.lightbend.java.modelServer.flinkprocessors;

import com.lightbend.java.modelServer.model.ServingResult;
import com.lightbend.java.modelServer.model.speculative.*;
import com.lightbend.java.modelServer.processor.VotingDecider;
import com.lightbend.speculative.Speculativedescriptor;
import org.apache.flink.api.common.state.ValueState;
import org.apache.flink.api.common.state.ValueStateDescriptor;
import org.apache.flink.api.common.typeinfo.TypeHint;
import org.apache.flink.api.common.typeinfo.TypeInformation;
import org.apache.flink.configuration.Configuration;
import org.apache.flink.streaming.api.functions.co.CoProcessFunction;
import org.apache.flink.util.Collector;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

public class SpeculativeProcessor extends CoProcessFunction<SpeculativeServiceRequest, Speculativedescriptor.SpeculativeDescriptor, ServingResult> {

    private static final long SERVERTIMEOUT = 150l;

    Decider speculativeDecider = new VotingDecider();

    // The managed keyed state see https://ci.apache.org/projects/flink/flink-docs-release-1.4/dev/stream/state.html
    ValueState<SpeculativeExecutionStats> specualtiveState;
    ValueState<Long> speculativeTimeout;

    // Because we are explicitely using unique GUIDs, we do not need to bring this to the key scope
    // Using this in the keygroup scope is good enough.
    Map<String, CurrentProcessing> currentProcessing = new HashMap<>();

    @Override public void open(Configuration parameters) {

        ValueStateDescriptor<SpeculativeExecutionStats> speculativeModelDesc = new ValueStateDescriptor<SpeculativeExecutionStats>(
                "currentSpeculativeModel",   // state name
                TypeInformation.of(new TypeHint<SpeculativeExecutionStats>() {})); // type information
        speculativeModelDesc.setQueryable("speculativeState");
        specualtiveState = getRuntimeContext().getState(speculativeModelDesc);
        ValueStateDescriptor<Long> speculativeTmoutDesc = new ValueStateDescriptor<Long> (
                "Timeout",
                TypeInformation.of(new TypeHint<Long>() {})); // type information
        speculativeTimeout = getRuntimeContext().getState(speculativeTmoutDesc);
    }


    @Override public void processElement1(SpeculativeServiceRequest record, Context ctx, Collector<ServingResult> out) throws Exception {

        if(specualtiveState.value() == null) specualtiveState.update(new SpeculativeExecutionStats(record.getDataType(), speculativeDecider.getClass().getName(), SERVERTIMEOUT));
        if(speculativeTimeout.value() == null) speculativeTimeout.update(SERVERTIMEOUT);

        // This will be invoked in 3 different use cases - process accordingly
        if(record.getResult().isPresent()){
            // This is either intermediate result from Model Processor or result from Router (no models)
            if (record.getModels() == 0){
                // from the router - just forward
                out.collect(record.getResult().get());
                return;
            }
            // Intermediate result
            CurrentProcessing results = currentProcessing.get(record.getGUID());
            if(results != null){
                // Still processing
                results.getResults().add(new ServingResponse(record.getGUID(), record.getResult().get()));
                if (results.getResults().size() >= results.getModels()){
                    // We are done
                    processResult(record.getGUID(), results, out);
                }
            }
            return;
        }
        // This is start collection
        // Add to the watch list
        long currentTime = System.currentTimeMillis();
        currentProcessing.put(record.getGUID(), new CurrentProcessing(record.getDataType(), record.getModels(), currentTime, new ArrayList<>()));
        // Schedule timeout
        ctx.timerService().registerProcessingTimeTimer(currentTime + speculativeTimeout.value());
     }

    @Override public void processElement2(Speculativedescriptor.SpeculativeDescriptor value, Context ctx, Collector<ServingResult> out) throws Exception {

        speculativeTimeout.update((value.getTmout()));
        if(specualtiveState.value() == null) specualtiveState.update(new SpeculativeExecutionStats(value.getDatatype(), speculativeDecider.getClass().getName(), value.getTmout()));
    }

    @Override public void onTimer(long timestamp, OnTimerContext ctx, Collector<ServingResult> out) {
        // Complete timed out requests
        try {

            long start = System.currentTimeMillis() - speculativeTimeout.value();
            for (Map.Entry entry : currentProcessing.entrySet()) {
                CurrentProcessing processing = (CurrentProcessing) entry.getValue();
                if (processing.getStart() < start) {
                    // Request timed out
                    System.out.println("Timed out request " + entry.getKey() + " at " + System.currentTimeMillis() + " started at " + processing.getStart());

                    processResult(entry.getKey().toString(), processing, out);
                }
            }
        }catch (Throwable t){}
    }

    // Complete speculative execution
    private void processResult(String GUID, CurrentProcessing results, Collector<ServingResult> out) throws Exception {
        // Run it through decider
        ServingResult servingResult = speculativeDecider.decideResult(results);
        // Return valye
        out.collect(servingResult);
        // Update state
        if(servingResult.isProcessed()) {
            SpeculativeExecutionStats state = specualtiveState.value();
            state.incrementUsage(servingResult.getDuration());
            specualtiveState.update(state);
        }
        // remove state
        currentProcessing.remove(GUID);
  }
}
